### PR TITLE
Implement border

### PR DESCRIPTION
**Pull request by hrsh7th**
#58 is good approach to support border for documentation.

But it is bit difficult because the implementation relates `vim-vital-vs` too.

So I implement it with referencing #58.

TODO and other concerns.

1. It should be an option?
2. border settings.
3. we should use `strdisplaywidth` for calculate positions.

https://user-images.githubusercontent.com/629908/103150434-891e8c80-47b7-11eb-8c7b-b8fffdf4d2e4.mp4

### CODE REVIEW

**Comment by hrsh7th** [comment]
https://github.com/hrsh7th/nvim-compe/issues/31

**Comment by tami5** [comment]
😍😍 Well done.

> It should be an option?

Definitely, like nvim_compe.border = true or { settings }
> border settings.

The color of border, transparency maybe 

> we should use strdisplaywidth for calculate positions.

What you mean? what are we using now

... other consideration

update buffer content if next content is not empty, otherwise close.

**Comment by hrsh7th** [comment]
Currently, vim-vital-vs's `s:FloatingWindow` will use a cached window.
But it remained the rendering garbage... So I choose this PR's approach (always close -> open)

File: autoload/compe/documentation.vim [hunk 1]
```diff
@@ -2,36 +2,59 @@ let s:MarkupContent = vital#compe#import('VS.LSP.MarkupContent')
...
-  \   'maxheight': float2nr(&lines * 0.4),
+  \   'row': l:layout.row + 1,
+  \   'col': l:layout.col + 1,
+  \   'maxwidth': l:maxwidth - 2,
+  \   'maxheight': l:maxheight - 2,
   \   'filetype': 'markdown',
   \   'contents': l:document,
+  \   'winhl': 'NormalFloat:Normal',
+  \ })
+
```
Comment (tami5) [inline autoload/compe/documentation.vim:34]: Add something like `chars = vim.g.compe.border.chars or [defaults]` to enable use customization.

Comment (hrsh7th) [reply autoload/compe/documentation.vim:34]: Good.

```
let g:compe = {}
let g:compe.border = v:false
let g:compe.border = {}
let g:compe.border.chars = {}
let g:compe.border.chars.top_left = ''
let g:compe.border.chars.top_right = ''
let g:compe.border.chars.bottom_left = ''
let g:compe.border.chars.bottom_right = ''
let g:compe.border.chars.vertical = ''
let g:compe.border.chars.horizontal = ''
```

I don't plan to add colors etc for now.

Comment (tami5) [reply autoload/compe/documentation.vim:34]: I'd say make it an array of chars [bottom_corner_l, bottom_corner_r, top_corner_r, top_corner_l, top/bottom, left/right]

so the user can do [char, char, ... ] in specific order.

**Comment by hrsh7th** [comment]
I will try it

**Comment by Ethaks** [comment]
[nvim_open_win()](https://neovim.io/doc/user/api.html#nvim_open_win()) now has an option `border`, allowing for `'single'`, `'double'`, and `'shadow'`, which makes the addition of borders as easy as adding `'border': 'single'` to the style, although that alone does not take care of a proper offset to account for the changed width etc. of course.
I only glanced at the MR dealing with borders however, so if the current idea already deals with all of this on its own, feel free to ignore my comment.

**Comment by cloggier** [comment]
Yes, using built in functionality should be the way to go here. It allows users to configure floating windows to look exactly the same.

**Comment by sam20908** [comment]
Looking forward to this getting merged! 🥳

**Comment by rcarriga** [comment]
For anyone else waiting for this, I've thrown together a quick branch with the new border in nightly: https://github.com/rcarriga/nvim-compe/tree/feat/doc-border. Obviously it's lacking any real customisation but if you're happy with defaults, it works!

**Comment by hrsh7th** [comment]
For now, the nvim nightly has a border option so we can provide border more easily...
@rcarriga Thank you for your information.

**Comment by sam20908** [comment]
I've tested the branch from rcarriga, and it works!

**Comment by folke** [comment]
Added support for borders to #426

**Pull request closed without merging**